lib/resty/core/socket.lua

```diff
@@ -107,17 +133,129 @@ local function setoption(cosocket, option, value)
                                                        err,
                                                        errlen)
     if rc ~= FFI_OK then
-        return nil, ffi_string(err, errlen[0])
+        return nil, ffi_str(err, errlen[0])
     end
 
     return true
 end
 
 
+local errmsg = base.get_errmsg_ptr()
+local session_ptr = ffi_new("void *[1]")
+local server_name_str = ffi_new("ngx_str_t[1]")
+local openssl_error_code = ffi_new("int[1]")
+
+
+local function setclientcert(cosocket, cert, pkey)
+    if not cert and not pkey then
+        cosocket.client_cert = nil
+        cosocket.client_pkey = nil
+        return
+    end
+
+    if not cert or not pkey then
+        error("client certificate must be supplied with corresponding " ..
+              "private key", 2)
+    end
+
+    if type(cert) ~= "cdata" then
+        error("bad client cert type", 2)
+    end
+
+    if type(pkey) ~= "cdata" then
+        error("bad client pkey type", 2)
+    end
+
+    cosocket.client_cert = cert
+    cosocket.client_pkey = pkey
+end
```

Comment:
Again, should we return `true` to indicate success?

Reply:
changed, thanks.